### ISSUE TITLE
Example for Parent and Child Windows

### ISSUE THREAD

**Issue by Delikt** [kojix2/LibUI#52]
Can we get a Example for HowTo Code a Main (Parent)Window and open a Child Window from it?

Maybe im to green behind my ears (Level 1 Developer in Praxis) but i dont get it to work.

The Problem is if i open a Window from the Main Window and Press X  (SIGKILL?) to close the Childwindow the whole Programm get closed  included my Parentwindow

Just one Example will be very helpful for me to get the Syntax :)

Thanks!

**Comment by AndyObtiva** [comment]
This is a known issue/limitation (though I just added to the [API Gotchas](https://github.com/AndyObtiva/glimmer-dsl-libui#api-gotchas) in Glimmer DSL for LibUI). Currently, if you spawn a window other than the main window, it is not treated as a child window, so when you close it, the entire application closes.

There is a [libui issue for supporting child windows](https://github.com/andlabs/libui/issues/137). It should address the issue once implemented (currently, libui improvements are happening in the [libui-ng](https://github.com/libui-ng/libui-ng) fork)

As for an example, here is one, but it suffers from the known issue mentioned above (it would not once child window support is added):

```ruby
require 'glimmer-dsl-libui'

include Glimmer

window('Main Window') {
  button('Spawn Child Window') {
    on_clicked do
      window('Child Window') {
        on_closing do
          puts 'Child window is closing'
        end
      }.show
    end
  }
  
  on_closing do
    puts 'Main window is closing'
  end
}.show
```

**Comment by cody271** [comment]
> This is a known issue/limitation (though I just added it to the [API Gotchas](https://github.com/AndyObtiva/glimmer-dsl-libui#api-gotchas) in Glimmer DSL for LibUI, built on top of LibUI). Currently, if you spawn a window other than the main window, it is not treated as a child window, so when you close it, the entire application closes.
> 
> There is a [libui issue for supporting child windows](https://github.com/andlabs/libui/issues/137). It should address the issue once implemented (currently, libui improvements are happening in the [libui-ng](https://github.com/libui-ng/libui-ng) fork)

I thought we had fixed this already, but perhaps not? Please [report](https://github.com/libui-ng/libui-ng/issues/new) any such issues you have upstream, thanks.

**Comment by AndyObtiva** [comment]
@cody271 I already mentioned that it’s reported as a libui issue:
https://github.com/andlabs/libui/issues/137

**Comment by cody271** [comment]
I'm not able to reproduce this issue. Maybe Ruby specific bug?

The C translation of the test case provided works fine:
```c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>

#include <ui.h>

uiWindow *mainwin;

static int onClosing(uiWindow *win, void *data)
{
	const char *msg = (char *)data;
	printf("%s\n", msg);

	if (win == mainwin)
		uiQuit();
	return 1;
}

static void spawnChildWindow(uiButton *btn, void *data)
{
	uiWindow *childwin = uiNewWindow("Child Window", 320, 240, 0);
	uiWindowOnClosing(childwin, onClosing, (void *)"Child Window is closing");
	uiControlShow(uiControl(childwin));
}

int main(void)
{
	uiInitOptions o;
	uiButton *btn;

	memset(&o, 0, sizeof (uiInitOptions));
	if (uiInit(&o) != NULL)
		abort();

	mainwin = uiNewWindow("Main Window", 320, 240, 0);
	btn = uiNewButton("Spawn Child Window");
	uiButtonOnClicked(btn, spawnChildWindow, NULL);
	uiWindowSetChild(mainwin, uiControl(btn));

	uiWindowOnClosing(mainwin, onClosing, (void *)"Main Window is closing");
	uiControlShow(uiControl(mainwin));
	uiMain();
	return 0;
}
```

**Comment by AndyObtiva** [comment]
@cody271 thank you for alerting me to this. I have some good news! Apparently, with one tiny modification, I can make the Ruby version work too:

```ruby
require 'glimmer-dsl-libui'

include Glimmer

window('Main Window') {
  button('Spawn Child Window') {
    on_clicked do
      window('Child Window') {
        on_closing do
          puts 'Child window is closing'
          1
        end
      }.show
    end
  }
  
  on_closing do
    puts 'Main window is closing'
  end
}.show
```

I just had the `on_closing` handler of the child window return `1` to indicate manually that it wants to close. Return values that are other than `1`/`0` would have resulted in the Glimmer DSL for LibUI default behavior of closing the window and then closing the app. 

I might modify that default behavior to autodetect if the window is not the last open window, and in which case, ensure that closing the window does not close the app. 

Otherwise, @Delikt this issue is truly resolved with the code included in this comment.

Cheers!

**Comment by cody271** [comment]
> I just had the `on_closing` handler of the child window return `1` to indicate manually that it wants to close. Return values that are other than `1`/`0` would have resulted in the Glimmer DSL for LibUI default behavior of closing the window and then closing the app.

Yup there are a few events that require a return value from the event handler, such as `uiWindowOnClosing()` and `uiOnShouldQuit()`.

**Comment by cody271** [comment]
> @cody271 I already mentioned that it’s reported as a libui issue:
> [andlabs/libui#137](https://github.com/andlabs/libui/issues/137)

Right I meant to the new repo, that old issue is from 2016 ..

You have a rather thorough list of outstanding issues in your [API Gotchas](https://github.com/AndyObtiva/glimmer-dsl-libui#api-gotchas) list!
If you'd like any of them addressed, feel free to open a new issue for them each. Thanks.

**Comment by AndyObtiva** [comment]
Hi,

I just wanted to note that I modified the default behavior in Glimmer DSL for LibUI to autodetect if the window being closed is a child window (not the main window), and if so, avoid quitting the app entirely, yet only close the child window, going back to the main window.

This was done in Glimmer DSL for LibUI v0.5.14:
https://github.com/AndyObtiva/glimmer-dsl-libui/tree/v0.5.14

I also included a new example called ["Basic Child Window"](https://github.com/AndyObtiva/glimmer-dsl-libui/tree/v0.5.14#basic-child-window), which is pretty much the code included above in a previous comment, but simplified to avoid returning `1` (Glimmer DSL for LibUI does not need the return value anymore to know to close the child window without quitting the app by default).

New code:

```ruby
require 'glimmer-dsl-libui'

include Glimmer

window('Main Window') {
  button('Spawn Child Window') {
    on_clicked do
      window('Child Window') {
        on_closing do
          puts 'Child window is closing'
        end
      }.show
    end
  }
  
  on_closing do
    puts 'Main window is closing'
  end
}.show
```

![screen1](https://raw.githubusercontent.com/AndyObtiva/glimmer-dsl-libui/v0.5.14/images/glimmer-dsl-libui-mac-basic-child-window.png)

![screen2](https://raw.githubusercontent.com/AndyObtiva/glimmer-dsl-libui/v0.5.14/images/glimmer-dsl-libui-mac-basic-child-window-open.png)

Cheers!
